File: LAB 2/rev4/SERVIDOR_tetris/tetris/TCPServer50.java
package tetris;

import java.net.ServerSocket;
import java.net.Socket;
import java.util.LinkedHashMap;
import java.util.Map;

public class TCPServer50 {
    public static final int DEFAULT_PORT = 4444;

    private final int serverPort;
    private final OnMessageReceived messageListener;
    private final Map<Integer, TCPServerThread50> clients = new LinkedHashMap<>();
    private volatile boolean running = false;
    private ServerSocket serverSocket;
    private int nextClientId = 1;

    public TCPServer50(OnMessageReceived messageListener) {
        this(DEFAULT_PORT, messageListener);
    }

    public TCPServer50(int serverPort, OnMessageReceived messageListener) {
        this.serverPort = serverPort;
        this.messageListener = messageListener;
    }

    public void run() {
        running = true;
        try {
            serverSocket = new ServerSocket(serverPort);
            System.out.println("Servidor TCP escuchando en el puerto " + serverPort);

            while (running) {
                Socket socket = serverSocket.accept();
                int clientId;
                TCPServerThread50 worker;

                synchronized (this) {
                    clientId = nextClientId++;
                    worker = new TCPServerThread50(socket, this, clientId);
                    clients.put(clientId, worker);
                }

                new Thread(worker, "Cliente-" + clientId).start();
                System.out.println("Nuevo jugador conectado: J" + clientId);
            }
        } catch (Exception e) {
            if (running) {
                System.out.println("Error en servidor TCP: " + e.getMessage());
            }
        } finally {
            stopServer();
        }
    }

    public synchronized void sendMessageTCPServer(String message) {
        for (TCPServerThread50 client : clients.values()) {
            client.sendMessage(message);
        }
    }

    public synchronized void sendToClient(int clientId, String message) {
        TCPServerThread50 client = clients.get(clientId);
        if (client != null) {
            client.sendMessage(message);
        }
    }

    public synchronized int getClientCount() {
        return clients.size();
    }

    public synchronized boolean hasClient(int clientId) {
        return clients.containsKey(clientId);
    }

    public synchronized void removeClient(int clientId) {
        clients.remove(clientId);
        if (messageListener != null) {
            messageListener.clientDisconnected(clientId);
        }
    }

    void notifyClientReady(int clientId) {
        if (messageListener != null) {
            messageListener.clientConnected(clientId);
        }
    }

    void notifyMessageReceived(int clientId, String message) {
        if (messageListener != null) {
            messageListener.messageReceived(clientId, message);
        }
    }

    public synchronized void stopServer() {
        running = false;
        for (TCPServerThread50 client : clients.values()) {
            client.stopClient();
        }
        clients.clear();
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (Exception ignored) {
        }
    }

    public interface OnMessageReceived {
        void messageReceived(int clientId, String message);
        default void clientConnected(int clientId) {}
        default void clientDisconnected(int clientId) {}
    }
}

File: LAB 4/Server50/app/src/main/java/com/example/server50/TCPServer50.kt
package com.example.server50

import android.util.Log
import java.net.ServerSocket
import java.net.Socket

class TCPServer50(
    private val portIp__: Int,
    private val messageListener: OnMessageReceived?
) {

    @Volatile
    private var running = false

    private var serverSocket: ServerSocket? = null
    private val maxClientes = 10
    private val sendclis = arrayOfNulls<TCPServerThread50>(maxClientes + 1)

    companion object {
        const val SERVERPORT = 8189
        private const val TAG = "TCPServer50"
    }

    fun getMessageListener(): OnMessageReceived? = messageListener

    fun run() {
        running = true

        try {
            Log.d(TAG, "TCP Server S: Connecting...")
            serverSocket = ServerSocket(portIp__)

            while (running) {
                val client: Socket = serverSocket!!.accept()
                val idDisponible = obtenerIdDisponible()

                if (idDisponible == -1) {
                    Log.e(TAG, "Servidor lleno. Cliente rechazado.")
                    try {
                        client.close()
                    } catch (e: Exception) {
                        Log.e(TAG, "Error cerrando cliente rechazado: ${e.message}")
                    }
                    continue
                }

                val ipCliente = client.inetAddress.hostAddress ?: "IP desconocida"
                val hiloCliente = TCPServerThread50(
                    client = client,
                    tcpserver = this,
                    clientID = idDisponible,
                    clientIP = ipCliente
                )

                registrarCliente(idDisponible, hiloCliente)
                Thread(hiloCliente).start()

                Log.d(TAG, "Cliente conectado en slot $idDisponible")
            }

        } catch (e: Exception) {
            if (running) {
                Log.e(TAG, "Error en servidor: ${e.message}", e)
            } else {
                Log.d(TAG, "Servidor detenido correctamente.")
            }
        } finally {
            cerrarTodosLosClientes()
            try {
                serverSocket?.close()
            } catch (e: Exception) {
                Log.e(TAG, "Error cerrando ServerSocket: ${e.message}")
            }
            serverSocket = null
            running = false
        }
    }

    @Synchronized
    private fun obtenerIdDisponible(): Int {
        for (i in 1..maxClientes) {
            if (sendclis[i] == null) return i
        }
        return -1
    }

    @Synchronized
    private fun registrarCliente(id: Int, cliente: TCPServerThread50) {
        sendclis[id] = cliente
    }

    @Synchronized
    fun liberarCliente(id: Int) {
        if (id in 1..maxClientes) {
            sendclis[id] = null
            Log.d(TAG, "ID liberado: Client $id")
        }
    }

    fun notificarMensajeCliente(clientID: Int, message: String) {
        Log.d(TAG, "Mensaje recibido de Client $clientID: $message")
        notifyUI(message)
        broadcastRaw(message)
    }

    fun notificarDesconexionCliente(clientID: Int) {
        Log.d(TAG, "Cliente desconectado en slot $clientID")
    }

    fun sendMessageTCPServer(message: String) {
        broadcastRaw(message)
    }

    @Synchronized
    private fun broadcastRaw(message: String) {
        for (i in 1..maxClientes) {
            val cliente = sendclis[i]
            if (cliente != null) {
                Thread {
                    cliente.sendMessage(message)
                    Log.d(TAG, "Mensaje reenviado a Client $i")
                }.start()
            }
        }
    }

    fun stop() {
        running = false
        try {
            serverSocket?.close()
        } catch (e: Exception) {
            Log.e(TAG, "Error cerrando servidor: ${e.message}")
        }
        cerrarTodosLosClientes()
    }

    @Synchronized
    private fun cerrarTodosLosClientes() {
        for (i in 1..maxClientes) {
            sendclis[i]?.stopClient()
            sendclis[i] = null
        }
    }

    private fun notifyUI(message: String) {
        messageListener?.messageReceived(message)
    }

    interface OnMessageReceived {
        fun messageReceived(message: String)
    }
}
